File: src/main/java/cn/academy/vanilla/teleporter/skills/FleshRipping.java
/**
 * Copyright (c) Lambda Innovation, 2013-2015
 * 本作品版权由Lambda Innovation所有。
 * http://www.li-dev.cn/
 *
 * This project is open-source, and it is distributed under
 * the terms of GNU General Public License. You can modify
 * and distribute freely as long as you follow the license.
 * 本项目是一个开源项目，且遵循GNU通用公共授权协议。
 * 在遵照该协议的情况下，您可以自由传播和修改。
 * http://www.gnu.org/licenses/gpl.html
 */
package cn.academy.vanilla.teleporter.skills;

import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.util.Vec3;
import cn.academy.ability.api.Skill;
import cn.academy.ability.api.ctrl.ActionManager;
import cn.academy.ability.api.ctrl.Cooldown;
import cn.academy.ability.api.ctrl.SkillInstance;
import cn.academy.ability.api.ctrl.action.SkillSyncAction;
import cn.academy.ability.api.data.AbilityData;
import cn.academy.core.client.ACRenderingHelper;
import cn.academy.core.client.sound.ACSounds;
import cn.academy.vanilla.teleporter.entity.EntityBloodSplash;
import cn.academy.vanilla.teleporter.entity.EntityMarker;
import cn.academy.vanilla.teleporter.util.TPAttackHelper;
import cn.lambdalib.util.generic.RandUtils;
import cn.lambdalib.util.generic.VecUtils;
import cn.lambdalib.util.helper.Color;
import cn.lambdalib.util.helper.Motion3D;
import cn.lambdalib.util.mc.EntitySelectors;
import cn.lambdalib.util.mc.Raytrace;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

/**
 * @author WeAthFolD
 */
public class FleshRipping extends Skill {

	public static final FleshRipping instance = new FleshRipping();

	private FleshRipping() {
		super("flesh_ripping", 3);
	}

	public static float getDamage(AbilityData data) {
		return instance.callFloatWithExp("damage", data);
	}

	public static float getRange(AbilityData data) {
		return instance.callFloatWithExp("range", data);
	}

	public static float getDisgustProb() {
		return instance.getFloat("disgust_prob");
	}

	@Override
	public SkillInstance createSkillInstance(EntityPlayer player) {
		return new SkillInstance().addChild(new FRAction()).setEstmCP(instance.getConsumption(AbilityData.get(player)));
	}

	public static class FRAction extends SkillSyncAction {

		public FRAction() {
			super(-1);
		}

		@Override
		public void onStart() {
			super.onStart();
			if (isRemote) {
				startEffects();
			}
		}

		@Override
		public void onTick() {
			if (isRemote) {
				updateEffects();
			} else {
				if (!cpData.canPerform(instance.getConsumption(aData)))
					ActionManager.abortAction(this);
			}
		}

		AttackTarget target;

		@Override
		public void writeNBTFinal(NBTTagCompound tag) {
			target = getAttackTarget();
			tag.setTag("x", target.toNBT());
		}

		@Override
		public void readNBTFinal(NBTTagCompound tag) {
			target = new AttackTarget((NBTTagCompound) tag.getTag("x"));
		}

		@Override
		public void onEnd() {
			if (target.target == null) {
				onAbort();
			} else {
				if (!isRemote) {
					cpData.performWithForce(instance.getOverload(aData), instance.getConsumption(aData));
					TPAttackHelper.attack(player, instance, target.target, getDamage(aData));
					if (RandUtils.ranged(0, 1) < getDisgustProb()) {
						player.addPotionEffect(new PotionEffect(Potion.confusion.id, 100));
					}

					setCooldown(instance, instance.getCooldown(aData));
					aData.addSkillExp(instance, instance.getFloat("expincr"));
				}
			}
		}

		@Override
		public void onAbort() {
			target = null;
		}

		@Override
		public void onFinalize() {
			if (isRemote) {
				endEffects();
			}
		}

		// CLIENT
		@SideOnly(Side.CLIENT)
		EntityMarker marker;

		static final Color DISABLED_COLOR = new Color().setColor4i(74, 74, 74, 160),
				THREATENING_COLOR = new Color().setColor4i(185, 25, 25, 180);

		@SideOnly(Side.CLIENT)
		private void startEffects() {
			if (isLocal()) {
				marker = new EntityMarker(player.worldObj);
				marker.setPosition(player.posX, player.posY, player.posZ);
				marker.color = DISABLED_COLOR;

				player.worldObj.spawnEntityInWorld(marker);
			}
		}

		@SideOnly(Side.CLIENT)
		private void updateEffects() {
			if (isLocal()) {
				AttackTarget at = getAttackTarget();
				marker.setPosition(at.dest.xCoord, at.dest.yCoord, at.dest.zCoord);
				if (at.target == null) {
					marker.color = DISABLED_COLOR;
					marker.width = marker.height = 1.0f;
				} else {
					marker.color = THREATENING_COLOR;
					marker.width = at.target.width * 1.2f;
					marker.height = at.target.height * 1.2f;
				}
			}
		}

		@SideOnly(Side.CLIENT)
		private void endEffects() {
			if (isLocal())
				marker.setDead();

			if (target != null && target.target != null) {
				ACSounds.playClient(player, "tp.guts", 0.6f);

				Entity e = target.target;
				int count = RandUtils.rangei(4, 6);
				while (count-- > 0) {
					double y = e.posY + RandUtils.ranged(0, 1) * e.height;
					if (e instanceof EntityPlayer)
						y += ACRenderingHelper.getHeightFix((EntityPlayer) e);

					double theta = RandUtils.ranged(0, Math.PI * 2);
					double r = 0.5 * RandUtils.ranged(0.8 * e.width, e.width);

					EntityBloodSplash splash = new EntityBloodSplash(player.worldObj);
					splash.setPosition(e.posX + r * Math.sin(theta), y, e.posZ + r * Math.cos(theta));
					player.worldObj.spawnEntityInWorld(splash);
				}
			}
		}

		private AttackTarget getAttackTarget() {
			double range = getRange(aData);
			MovingObjectPosition trace = Raytrace.traceLiving(player, range, EntitySelectors.living);

			Entity target = null;
			Vec3 dest;

			if (trace != null) {
				target = trace.entityHit;
				dest = trace.hitVec;
			} else {
				dest = new Motion3D(player, true).move(range).getPosVec();
			}

			return new AttackTarget(dest, target);
		}

		class AttackTarget {
			public final Vec3 dest;
			public final Entity target;

			public AttackTarget(Vec3 _dest, Entity _target) {
				dest = _dest;
				target = _target;
			}

			public AttackTarget(NBTTagCompound tag) {
				this(VecUtils.vec(tag.getFloat("x"), tag.getFloat("y"), tag.getFloat("z")),
						player.worldObj.getEntityByID(tag.getInteger("i")));
			}

			NBTTagCompound toNBT() {
				NBTTagCompound ret = new NBTTagCompound();
				ret.setFloat("x", (float) dest.xCoord);
				ret.setFloat("y", (float) dest.yCoord);
				ret.setFloat("z", (float) dest.zCoord);
				ret.setInteger("i", target == null ? 0 : target.getEntityId());

				return ret;
			}

		}

	}
}

File: src/main/scala/cn/academy/vanilla/teleporter/skill/FleshRipping.scala
/**
  * Copyright (c) Lambda Innovation, 2013-2016
  * This file is part of the AcademyCraft mod.
  * https://github.com/LambdaInnovation/AcademyCraft
  * Licensed under GPLv3, see project root for more information.
  */
package cn.academy.vanilla.teleporter.skill

import cn.academy.ability.api.Skill
import cn.academy.ability.api.context.{ClientContext, ClientRuntime, Context, RegClientContext}
import cn.academy.core.client.ACRenderingHelper
import cn.academy.core.client.sound.ACSounds
import cn.academy.vanilla.generic.entity.EntityBloodSplash
import cn.academy.vanilla.teleporter.entity.EntityMarker
import cn.academy.vanilla.teleporter.util.TPSkillHelper
import cn.lambdalib.annoreg.core.Registrant
import cn.lambdalib.s11n.{SerializeIncluded, SerializeNullable}
import cn.lambdalib.s11n.network.NetworkMessage.Listener
import cn.lambdalib.s11n.network.NetworkS11n.NetworkS11nType
import cn.lambdalib.util.generic.{RandUtils, VecUtils}
import cn.lambdalib.util.helper.{Color, Motion3D}
import cn.lambdalib.util.mc.{EntitySelectors, Raytrace}
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.potion.{Potion, PotionEffect}
import net.minecraft.util.{MovingObjectPosition, Vec3}

/**
  * @author WeAthFolD, KSkun
  */
object FleshRipping extends Skill("flesh_ripping", 3) {

  @SideOnly(Side.CLIENT)
  override def activate(rt: ClientRuntime, keyID: Int) = activateSingleKey(rt, keyID, p => new FRContext(p))

}

object FRContext {

  final val MSG_ABORT = "abort"
  final val MSG_END = "end"
  final val MSG_EFFECT_END = "effect_end"

}

import cn.academy.ability.api.AbilityAPIExt._
import cn.lambdalib.util.generic.MathUtils._
import FRContext._

class FRContext(p: EntityPlayer) extends Context(p, FleshRipping) {

  private val exp = ctx.getSkillExp

  private var target: AttackTarget = _

  @Listener(channel=MSG_TICK, side=Array(Side.SERVER))
  private def s_tick() = {
    if(!ctx.canConsumeCP(getConsumption)) terminate()
    target = getAttackTarget
  }

  @Listener(channel=MSG_END, side=Array(Side.SERVER))
  private def s_end() = {
    sendToClient(MSG_EFFECT_END, target)
    if(target.target == null) sendToSelf(MSG_ABORT)
    else {
      ctx.consumeWithForce(getOverload, getConsumption)
      TPSkillHelper.attack(ctx, target.target, getDamage)
      if(RandUtils.ranged(0, 1) < getDisgustProb) player.addPotionEffect(new PotionEffect(Potion.confusion.id, 100))
      ctx.setCooldown(lerpf(100, 60, exp).toInt)
      ctx.addSkillExp(.005f)
    }
    terminate()
  }

  @Listener(channel=MSG_KEYUP, side=Array(Side.CLIENT))
  private def l_onKeyUp = sendToServer(MSG_END)

  @Listener(channel=MSG_KEYABORT, side=Array(Side.CLIENT))
  private def l_onKeyAbort = {
    sendToSelf(MSG_EFFECT_END, new AttackTarget())
    terminate()
  }

  @Listener(channel=MSG_ABORT, side=Array(Side.SERVER))
  private def s_abort = target = null

  private def getDamage: Float = lerpf(9, 15, exp)

  override def getRange: Double = lerpf(6, 14, exp)

  private def getDisgustProb: Float = .05f

  private def getConsumption: Float = lerpf(247, 302, exp)

  private def getOverload: Float = lerpf(60, 50, exp)

  def getAttackTarget: AttackTarget = {
    val range: Double = getRange
    val trace: MovingObjectPosition = Raytrace.traceLiving(player, range, EntitySelectors.living)
    var target: Entity = null
    var dest: Vec3 = null
    if (trace != null) {
      target = trace.entityHit
      dest = trace.hitVec
    }
    else dest = new Motion3D(player, true).move(range).getPosVec
    new AttackTarget(dest, target, player)
  }

}

@Registrant
@SideOnly(Side.CLIENT)
@RegClientContext(classOf[FRContext])
class FRContextC(par: FRContext) extends ClientContext(par) {

  private var marker: EntityMarker = _

  private val DISABLED_COLOR: Color = new Color().setColor4i(74, 74, 74, 160)
  private val THREATENING_COLOR: Color = new Color().setColor4i(185, 25, 25, 180)

  @Listener(channel=MSG_TERMINATED, side=Array(Side.CLIENT))
  private def l_terminated() = {
    if(marker != null) marker.setDead()
  }

  @Listener(channel=MSG_MADEALIVE, side=Array(Side.CLIENT))
  private def l_startEffect() = {
    if(isLocal) {
      marker = new EntityMarker(player.worldObj)
      marker.setPosition(player.posX, player.posY, player.posZ)
      marker.color = DISABLED_COLOR
      player.worldObj.spawnEntityInWorld(marker)
    }
  }

  @Listener(channel=MSG_TICK, side=Array(Side.CLIENT))
  private def l_updateEffect() = {
    if(isLocal) {
      val at: AttackTarget = par.getAttackTarget
      marker.setPosition(at.dest.xCoord, at.dest.yCoord, at.dest.zCoord)
      if (at.target == null) {
        marker.color = DISABLED_COLOR
        marker.width = 1.0f
        marker.height = 1.0f
      }
      else {
        marker.color = THREATENING_COLOR
        marker.width = at.target.width * 1.2f
        marker.height = at.target.height * 1.2f
      }
    }
  }

  @Listener(channel=MSG_EFFECT_END, side=Array(Side.CLIENT))
  private def c_endEffect(target: AttackTarget) = {
    if(isLocal) marker.setDead()

    if(target != null && target.target != null) {
      ACSounds.playClient(player, "tp.guts", 0.6f)
      val e: Entity = target.target
      for(i <- 0 to RandUtils.rangei(4, 6)) {
        var y: Double = e.posY + RandUtils.ranged(0, 1) * e.height
        e match {
          case player1: EntityPlayer => y += ACRenderingHelper.getHeightFix(player1)
          case _ =>
        }
        val theta: Double = RandUtils.ranged(0, Math.PI * 2)
        val r: Double = 0.5 * RandUtils.ranged(0.8 * e.width, e.width)
        val splash: EntityBloodSplash = new EntityBloodSplash(player.worldObj)
        splash.setPosition(e.posX + r * Math.sin(theta), y, e.posZ + r * Math.cos(theta))
        player.worldObj.spawnEntityInWorld(splash)
      }
    }
  }

}

@Registrant
@NetworkS11nType
class AttackTarget() {

  @SerializeIncluded
  var dest: Vec3 = _
  @SerializeIncluded
  @SerializeNullable
  var target: Entity = _
  @SerializeIncluded
  var player: EntityPlayer = _

  def this(_dest: Vec3, _target: Entity, _player: EntityPlayer) {
    this()
    dest = _dest
    target = _target
    player = _player
  }

  def this(tag: NBTTagCompound, _player: EntityPlayer) {
    this(VecUtils.vec(tag.getFloat("x"), tag.getFloat("y"), tag.getFloat("z")),
      _player.worldObj.getEntityByID(tag.getInteger("i")), _player)
  }

  private def toNBT: NBTTagCompound = {
    val ret: NBTTagCompound = new NBTTagCompound
    ret.setFloat("x", dest.xCoord.toFloat)
    ret.setFloat("y", dest.yCoord.toFloat)
    ret.setFloat("z", dest.zCoord.toFloat)
    ret.setInteger("i", if(target == null) 0 else target.getEntityId)
    ret
  }

}
